lib/collection/src/shards/resharding/mod.rs

```diff
@@ -1,13 +1,52 @@
+use std::fmt;
+
 use schemars::JsonSchema;
 use segment::types::ShardKey;
 use serde::{Deserialize, Serialize};
 
 use super::shard::{PeerId, ShardId};
 
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct ReshardingState {
+    pub peer_id: PeerId,
+    pub shard_id: ShardId,
+    pub shard_key: Option<ShardKey>,
+}
+
+impl ReshardingState {
+    pub fn new(peer_id: PeerId, shard_id: ShardId, shard_key: Option<ShardKey>) -> Self {
+        Self {
+            peer_id,
+            shard_id,
+            shard_key,
+        }
+    }
+
+    pub fn matches(&self, key: &ReshardingKey) -> bool {
+        self.peer_id == key.peer_id
+            && self.shard_id == key.shard_id
+            && self.shard_key == key.shard_key
```

Comment:
How about `self.key() == key`?

```suggestion
        self.key() == key
```

Reply:
I don't like the extra clone. 😂